packages/trip-form/src/MetroModeSelector/index.tsx

```diff
@@ -139,76 +194,89 @@ function ModeButton({
+  const interactionProps = getReferenceProps();
+
+  // ARIA roles are added by the `useRole` hook.
+  // Remove the aria-controls, aria-expanded, and aria-haspopup props from the label, they will
+  // instead be passed to the button for keyboard/screen reader users to trigger the popup.
+  const {
+    "aria-controls": ariaControls,
+    "aria-expanded": ariaExpanded,
+    "aria-haspopup": ariaHasPopup,
+    ...labelInteractionProps
+  } = interactionProps;
 
   return (
-    <>
-      {/* useRole adds aria-controls */}
-      {/* eslint-disable-next-line jsx-a11y/role-supports-aria-props */}
-      <ModeButtonItem
-        className={modeButton.enabled ? "enabled" : ""}
-        ref={reference}
-        role="checkbox"
+    <ModeButtonWrapper>
+      {/* Basic checkbox that states whether a mode is selected. */}
+      <input
+        checked={modeButton.enabled ?? undefined}
+        id={`metro-mode-selector-mode-${id}`}
+        onChange={onToggle}
+        type="checkbox"
+      />
+      {/* Label for the above checkbox, placed right after, so that CSS applies based on checkbox state. */}
+      {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
+      <label
         // This library relies on prop spreading
-        /* eslint-disable-next-line react/jsx-props-no-spreading */
-        {...getReferenceProps()}
-        aria-expanded={renderDropdown}
-        aria-checked={modeButton.enabled ?? false}
-        aria-label={modeButton.label}
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...labelInteractionProps}
+        htmlFor={`metro-mode-selector-mode-${id}`}
+        // This will trigger mouse effects such as showing popup on hover of on check.
+        ref={reference}
         title={modeButton.label}
-        // Defaults to true
-        fillModeIcons={fillModeIcons !== false}
       >
-        <modeButton.Icon size={32} />
-      </ModeButtonItem>
+        <modeButton.Icon role="none" size={32} />
+        <InvisibleA11yLabel>{modeButton.label}</InvisibleA11yLabel>
+      </label>
+      <button
+        // eslint-disable-next-line react/jsx-props-no-spreading
+        {...interactionProps}
+        // Disable button if mode is not checked (but keep in DOM for screen reader awareness)
+        disabled={!modeButton.enabled}
+        // Required by linter settings
+        type="button"
+      >
+        <span role="none">
+          {open ? <CaretUp size={14} /> : <CaretDown size={14} />}
+        </span>
+        <InvisibleA11yLabel>
+          <FormattedMessage
+            defaultMessage={defaultMessages["otpUi.ModeSelector.settingsLabel"]}
+            description="Label for the button to open settings for a travel mode."
+            id="otpUi.ModeSelector.settingsLabel"
+            values={{ mode: modeButton.label }}
+          />
+        </InvisibleA11yLabel>
+      </button>
       {renderDropdown && (
         <FloatingFocusManager context={context} modal={false}>
```

Comment:
Adjusting `FloatingFocusManager` like this fixes the issue of the buttons gaining focus when the popup closes, and it also traps the tab focus inside the popup until `esc` is pressed. We could also add a dismiss button that is hidden visually but can be tabbed to, if `esc` is not good enough.

```suggestion
        <FloatingFocusManager context={context} returnFocus={false}>
```

Reply:
Good point on keeping the focus within the popup. Maybe it should just be `<FloatingFocusManager context={context}>`, as we want to return the focus to the parent button when dismissing the popup.

Reply:
See https://github.com/ibi-group/otp-ui/pull/13/commits/c80bbdd2cca1af487502964e9f2442169c753ebf for updates. Feel free to merge into your PR branch when you are ready and have no other comments.

Reply:
Do we need it to return the focus? The focus ring appearing when you leave the hover element could be confusing for folks. I tested it with keyboard with `returnFocus=false` and it seemed like focus still went back to where it should, but maybe I missed a use case?

Reply:
Slight difference: it doesn't refocus the mode selector upon hitting `esc`, but when you press tab again it starts from the correct spot.

Reply:
If after reading this you still believe we should have returnFocus as true, please merge this! No need for more discussion.

Reply:
I think we should **show**, visually, a focused element when dismissing the popup using ESC. That way keyboard users know where they are. It looks like with `returnFocus={false}`, when dismissing the popup, the focus might remain within the `<FloatingFocusManager>`, which contents would be invisible.